src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SemanticFacts/CSharpSemanticFacts.cs

```diff
@@ -57,23 +57,51 @@ public bool CanReplaceWithRValue(SemanticModel semanticModel, [NotNullWhen(true)
         {
             var location = token.GetLocation();
 
-            foreach (var ancestor in token.GetAncestors<SyntaxNode>())
+            var tokenParent = token.Parent;
+            var sourceAncestor = tokenParent;
+            // Skip the tuple expression if the identifier implies the name of the tuple field
+            if (tokenParent is IdentifierNameSyntax { Parent: ArgumentSyntax { Parent: TupleExpressionSyntax parentTuple } })
+            {
+                sourceAncestor = parentTuple.Parent;
+            }
+
+            if (sourceAncestor is null)
+                return null;
+
+            foreach (var ancestor in sourceAncestor.AncestorsAndSelf())
             {
                 var symbol = semanticModel.GetDeclaredSymbol(ancestor, cancellationToken);
                 if (symbol != null)
                 {
-                    if (symbol is IMethodSymbol { MethodKind: MethodKind.Conversion })
-                    {
-                        // The token may be part of a larger name (for example, `int` in `public static operator int[](Goo g);`.
-                        // So check if the symbol's location encompasses the span of the token we're asking about.
-                        if (symbol.Locations.Any(static (loc, location) => loc.SourceTree == location.SourceTree && loc.SourceSpan.Contains(location.SourceSpan), location))
-                            return symbol;
-                    }
-                    else
+                    switch (symbol)
                     {
-                        // For any other symbols, we only care if the name directly matches the span of the token
-                        if (symbol.Locations.Contains(location))
-                            return symbol;
+                        case IMethodSymbol { MethodKind: MethodKind.Conversion }:
+                            {
+                                // The token may be part of a larger name (for example, `int` in `public static operator int[](Goo g);`.
+                                // So check if the symbol's location encompasses the span of the token we're asking about.
+                                if (symbol.Locations.Any(static (loc, location) => loc.SourceTree == location.SourceTree && loc.SourceSpan.Contains(location.SourceSpan), location))
+                                    return symbol;
+                                break;
+                            }
+
+                        case IPropertySymbol { ContainingType: ITypeSymbol { IsAnonymousType: true } }:
+                            {
+                                Debug.Assert(ancestor is AnonymousObjectMemberDeclaratorSyntax);
```

Comment:
the assert doesn't help.  we're going to crash on the next line if this is false.

also, we might as well just do this as a normal typecheck of ancestor, instead of a blind cast.